src/main/java/htmlpublisher/util/FileVisitorService.java

```diff
@@ -0,0 +1,166 @@
+package htmlpublisher.util;
+
+import hudson.util.FileVisitor;
+
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.HashSet;
+import java.util.Set;
+
+import java.io.File;
+import java.io.IOException;
+
+/**
+ * Manages file visits and dispatches them over multiple visit workers
+ */
+public class FileVisitorService implements AutoCloseable {
+
+	/**
+	 * Marker for stopping the worker
+	 */
+	private static final VisitEntry POISON_PILL = new VisitEntry(null, null);
+
+	/**
+	 * Origin visitor that does the "real" work
+	 */
+	private FileVisitor delegateVisitor;
+
+	/**
+	 * Files, that have been collected via directory scan
+	 */
+	BlockingQueue<VisitEntry> queue = new LinkedBlockingQueue<>();
+
+	/**
+	 * Workers, that delegates the collected files
+	 */
+	private Set<Future<IOException>> workers = new HashSet<>();
+
+	/**
+	 * Indicates that the visitor service is currently shutting down -> no more
+	 * collects allowed
+	 */
+	private boolean shutdown = false;
+
+	/**
+	 * Exception that has been thrown while doing the work
+	 */
+	private IOException exception;
+
+	static class VisitEntry {
+
+		private File file;
+		private String relativePath;
+
+		VisitEntry(File file, String relativePath) {
+			this.file = file;
+			this.relativePath = relativePath;
+		}
+
+	}
+
+	class Worker implements Callable<IOException> {
+
+		@Override
+		public IOException call() {
+
+			try {
+				while (true) {

```

Comment:
I don't fully understand here.
shouldn't create task for all files to copy then wait for the end of the tasks?